### ISSUE TITLE
 await syntax in sftp.get PassThrough() does not complete as expected

### ISSUE THREAD

**Issue by Dev10000** [theophilusx/ssh2-sftp-client#538]
The operation hangs when using the `await` syntax with the `sftp.get` method. However, the `.then` syntax works correctly. This behavior suggests a possible bug or issue with how the `sftp.get` method handles streams when used with `await`.

**Code to Reproduce**:
```typescript
import Client from "ssh2-sftp-client";
import { S3 } from "@aws-sdk/client-s3";
import { Upload } from "@aws-sdk/lib-storage";
import { PassThrough } from "stream";

async function streamFTPToCloudflare(): Promise<void> {
  const sftp = new Client();
  const s3 = new S3({
    endpoint: CLOUDFLARE_ENDPOINT_EU, // Cloudflare R2 endpoint
    region: "auto",
    credentials: {
      accessKeyId: CLOUDFLARE_KEY,
      secretAccessKey: CLOUDFLARE_SECRET,
    },
  });

  try {
    await sftp.connect({
      host: FTP_HOST,
      port: 22,
      username: FTP_USER,
      password: FTP_PASSWORD,
    });

    const passThrough = new PassThrough();

    // await sftp.get hangs
    console.log("BEFORE SFTP")
    await sftp.get('/pricelist-5.txt.zip', passThrough);
    console.log("AFTER SFTP")

    // .then syntax works
    // sftp
    // .get("/pricelist-5.txt.zip", passThrough)
    // .then(() => {
    //   console.log("File has been piped to PassThrough stream.");
    // })
    // .catch((err) => {
    //   console.error("Error in getting the file:", err);
    //   throw err;
    // });

    // Upload to Cloudflare using the PassThrough stream
    const uploadParams = {
      Bucket: "pim",
      Key: "uploaded-file-test.txt.zip",
      Body: passThrough,
    };

    const upload = new Upload({
      client: s3,
      params: uploadParams,
      leavePartsOnError: false, // optional manually handle dropped parts
    });

    upload.on("httpUploadProgress", (progress) => {
      console.log(progress);
    });

    await upload.done();
    console.log("Upload to Cloudflare R2 completed successfully.");
  } catch (error) {
    console.error("Error during SFTP to Cloudflare R2 transfer:", error);
    throw error;
  } finally {
    await sftp.end();
  }
}

streamFTPToCloudflare();
```

**Environment**:
- Node.js version: 20.14.0
- "ssh2-sftp-client": "^10.0.3"
- "typescript": "^5.4.5"
- "@aws-sdk/client-s3": "^3.596.0",
- "@aws-sdk/lib-storage": "^3.596.0",
- Operating system: Ubuntu 22.04.4 LTS

**Logs and Output**:
```
BEFORE SFTP
// Program hangs here and does not log "AFTER SFTP"
```

**Comment by theophilusx** [comment]
I'm not sure your conclusion follows. The ssh2-sftp-client code has no
knowledge about how it is called i.e. it doesn't know and does not care if it is being
called with await or is just a promise chain. The function just returns
a promise which is either resolved or rejected promise. The only
difference between using await and a promise train is how you handle the
returned promise. 

The first thing I would look at is the use of the pass through stream. I
would start by seeing if you can reproduce the issue without S3 and then
verify the S3 related code works without using ssh2-sftp-client. 

I would also add a debug function to the connect config object and see
exactly what is being logged by ssh2 and ssh2-sftp-client. The system
just 'hanging' is very rare - at the very least, the sftp connection
should timeout or be closed by the remote sftp server at some point. 

The 'unknown' in this scenario is s3 and
cloudFlare. For me to track this down any further, it would need to be
reproducible outside of using s3/cloudFlare (and typscript for that
matter. While I don't suspect it is an issue, there have bene issues in
th past due to problems with typescript). It is possible there is an
issue with handling of the pass through stream, though I cannot see how
that would differ between the function being called with await or as a
promise chain.

**Comment by STDigitalIN** [comment]
Same this is happening with me as well but with sftp.connect, sometimes it hangs without throwing any error/exception.

**Comment by Dev10000** [comment]
TLDR: In my case, the easiest solution was to remove the `await` keyword as it's unnecessary when passing through the stream. Since I don't write directly to a writable but need a readable stream from `sftp.get`, I can use `PassThrough` without `await` or, if I need to use `await`, I can use a `buffer`. Also, I can't see any scenario where you would need to use `await` directly in the `sftp.get` call anyway, when outputting a readable stream.

@theophilusx Thank you for the feedback. I did some more testing with minimal setup, and the same issue persists. I added the debug function and removed the Cloudflare object storage, using only `fs.createWriteStream`. With the debugger, I noticed that the stream in `PassThrough` starts flowing but then either interrupts or doesn't finish. When using `await` directly on sftp.get, the `await` keyword prevents it from being consumed correctly.

Also, the `.then `syntax didn't actually work correctly. Even though it seemed to work as mentioned in my first message, it didn't wait for the promise to resolve as it should. This is why it appeared to work similarly when removing the await keyword.

The issue seems to be more about the behaviour of the `PassThrough` duplex stream and how streams are consumed in Node.js rather than a specific limitation of `sftp.get` itself. Essentially, when using `await` directly with `sftp.get` and `PassThrough`, the promise does not resolve without adding an event listener `.on('data')` to the `PassThrough` stream. With `fs.createWriteStream` or other normal streams, this works because they are consumed directly. When adding a "data" listener, it doesn't hang when using `await`, but some reason it still doesn't write the file correctly, it writes the file, but it is 0 bytes (see the example). Therefore, if I need to use `await` directly in `sftp.get`, I need to use a buffer.

### Examples
Here are some of my test example snippets for cases when you can't directly use a writable as the destination in `sftp.get`, such as when writing to cloud storage or creating ETL pipelines etc.

1. **Remove await**
By just removing the `await` from sftp.get, PassThrough works with promises. With the `await` keyword, it seems that the stream is not consumed correctly and the flow interrupts.
```js
const pipelinePromise = promisify(pipeline);
const passThrough = new PassThrough();
// await sftp.get('/pricelist-5.txt.zip', passThrough); doesn't work
sftp.get('/pricelist-5.txt.zip', passThrough);
const writeStream = fs.createWriteStream('downloaded-file.zip');
await pipelinePromise(passThrough, writeStream);
```

2. **Different execution order `await` works**
```js
const passThrough = new PassThrough();
const writeStream = fs.createWriteStream('downloaded-file.zip');
passThrough.pipe(writeStream);
await sftp.get('/pricelist-5.txt.zip', passThrough);
//--------------------------------------------------
// This execution order doesn't work
const passThrough = new PassThrough();
await sftp.get('/pricelist-5.txt.zip', passThrough);
const writeStream = fs.createWriteStream('downloaded-file.zip');
passThrough.pipe(writeStream);
```
3. **Use Buffer with `await`**
If for some reason you need to use `await` directly in `sftp.get`, you can use a buffer.
```ts
const buffer = await sftp.get('/pricelist-5.txt.zip') as Buffer;
const writeStream = fs.createWriteStream('downloaded-file.zip');
const stream = Readable.from(buffer);
stream.pipe(writeStream);
```
4. **Attach Data Event Listener to Fix PassThrough Consumption**
When adding a data listener before `await sftp.get`, the stream should be consumed correctly but it still doesn't write to the disk correctly. Now it writes the file, but it is 0 bytes. Not sure why this doesn't work.
```js
const pipelinePromise = promisify(pipeline);
const passThrough = new PassThrough();
passThrough.on('data', (chunk) => {
    console.log(`PassThrough stream data chunk: ${chunk.length} bytes`);
});
await sftp.get('/pricelist-5.txt.zip', passThrough);
const writeStream = fs.createWriteStream('downloaded-file.zip');
await pipelinePromise(passThrough, writeStream);
```

**Comment by theophilusx** [comment]
Thanks for the info.

Will try to find time to dig deeper into this. I suspect it is related
to the consumption side - the stream the data is being fed into must be
consumed as ssh2-sftp-client is passing data through via pipe. Without
consumption on the far end, transfer will stall due to back pressure in
the pipe.

The fact it down't work with either await or promise chain make a lot
more sense. There is no real difference between await and promise chians
- it is just syntactic sugar. In the case where the destination for
get() is a stream, the promise will be rejected if the destination
stream or source stream throws an error and will be resolved succesfully
if the source stream generates an end event. The source stream will
generate an end event when all the data has been piped to the
destination stream or the source stream receives a close signal.

If the destination stream is not consuming the data, it will never reach
an end and therefore will not resolve successfully. As there are no
errors, it won't be rejected either. It will appear to hang until
something happens, such as closing the destination stream.

A pass through stream must ahve a consumer at the far end. A pass
through stream on its own wo't do anything. I've not looked at your
examples/tests in any detail yet, but I'm guessing the issue will turn
out to be a problem with whatever is consuming data from your pass
through stream. I've rarely used pass through streams and when I have,
it ahs been via a 3rd party package which handles the setup/config of
the stream. In cases like yours, I would just pass the write stream from
the S3 storage straight to get() - it isn't yet clear to me why you need
the passthrough stream. All that get() really does is
source.pipe(destination). Destination just has to be a stream which
pipe() can write to.

I'm concerned your solution to just not use await or promise chain is
going to cause you problems. For exmaple, you won't know when the
process has completed and therefore you won't know when it is OK to
close the connection. It is still very weird that you find it works
without await or promise chain - having an await or a then() makes
absolutely no difference to what goes on inside get(). As mentioned
before, get() has no knowledge and doe snot care how it was
called. Regardless of whether you call it with await or as a promise
chain, it creates a promise and that promise is either rejected or
fulfilled, regardless of whether you look at it with either await or
then().

**Comment by Dev10000** [comment]
> Thanks for the info. Will try to find time to dig deeper into this. I suspect it is related to the consumption side

I think `get()` is working correctly, it’s just the nature of how streams work. And in my specific case, I should not use `await` directly on `get()` with `PassThrough`. Ideally, it would be great if `get()` could return the `Readable`, such as `rdr = this.sftp.createReadStream(remotePath, options.readStreamOptions)`, but I can also achieve the same with `PassThrough`.

> The fact it down't work with either await or promise chain make a lot more sense. There is no real difference between await and promise chians - it is just syntactic sugar.

Yes, I know `await` and promise chains are just syntactic sugar, but for some reason, they behave a little differently in this case. `await` will hang, waiting for the promise to resolve, while `.then` doesn’t hang, and for some reason it doesn’t wait for the promise to resolve. `.then` should also wait for the promise to resolve. This difference is still a mystery to me.

> If the destination stream is not consuming the data, it will never reach an end and therefore will not resolve successfully. As there are no errors, it won't be rejected either. It will appear to hang until something happens, such as closing the destination stream. A pass through stream must ahve a consumer at the far end. A pass through stream on its own wo't do anything.

Yes, I noticed this. That's why if you make a pipeline before the `get()` method the `await` and promise chain works. I saw some of your old examples from previous issues (https://github.com/theophilusx/ssh2-sftp-client/issues/202#issuecomment-573252641) and noticed that you are creating the pipeline before calling `get()`. That’s why `await` or the promise chain works directly on the `get()` method, but I’m doing it differently by creating the pipeline after calling `get()`.

> In cases like yours, I would just pass the write stream from the S3 storage straight to get() - it isn't yet clear to me why you need the passthrough stream. All that get() really does is source.pipe(destination). Destination just has to be a stream which pipe() can write to.

I can't use `Write stream` with `S3 Upload`, as it requires a `Readable`. `S3 Upload` client takes an object as a parameter and requires either a `Buffer` or `Readable`. As I want to use `Readable` and `get()` doesn’t return Readable directly, I need to use `PassThrough` duplex stream that can be used as both `Writable` and `Readable`. Also, if you want to do ETL pipelines after the `get()`, you ideally need a `Readable` that you can pipe forward.

> I'm concerned your solution to just not use await or promise chain is going to cause you problems. For exmaple, you won't know when the process has completed and therefore you won't know when it is OK to close the connection. It is still very weird that you find it works without await or promise chain.

Please note that I am still using `await` but not directly on the `get()` method. I use `await` and resolve all promises later in the `promisified pipeline`, so the Node.js build-in `pipeline` method handles all the events automatically and ensures that streams are properly closed when they complete or when an error occurs. I believe S3 Upload is also promise-based so should work with that as well. I don’t think I need to wait for the promise to resolve directly on the `get()` method as long as the stream is handled correctly, for example with a `promisified pipeline` or `S3 Upload`.

If you call `pipeline` after the `get()` method, using `await` on `get()` would load the whole stream into memory, similarly to using a Buffer, which is not desirable when you plan to continue the stream. 

```js
const pipelinePromise = promisify(pipeline);
const passThrough = new PassThrough();
sftp.get('/pricelist-5.txt.zip', passThrough); // Don't use await here
const writeStream = fs.createWriteStream('downloaded-file.zip');
await pipelinePromise(passThrough, writeStream); // Here is the await and the promisfied pipeline that should handle all the events automatically and ensures that streams are properly closed when they complete or when an error occurs. 
```

> [!NOTE]
I believe, the `get()` implementation of promise resolving only at the end of stream and not returning/resolving to Redable could cause issues if you are not able to use promisified destination, then it would be safer to use `await` directly on `get()` and use Buffer or you would need to do the pipeline before calling the `get()`.

**Comment by theophilusx** [comment]
> Ideally, it would be great if get() could return the Readable, such as rdr =
> this.sftp.createReadStream(remotePath, options.readStreamOptions), but I can also achieve the same with
> PassThrough.

Return the read stream from get() causes issues for a Promise based
API. i.e. how would you determine when to resolve the promise in the
general case.

However, note that there is the createReadStream() method, so yo
could just use that and implement your own version of get(), wrapping it
in a promise if you want and resolving/rejecting the promise on whatever
criteria meets your requirements. In fact, this is sort of the reason I
added the createReadStream and createWriteStream methods as part of the
ssh2-sftp-client API. They provide some low level access, but avoid the
more messy aspects of directly using the underlying ssh2 library.
